cnd/src/btsieve/bitcoin/mod.rs

```diff
@@ -8,142 +8,228 @@ pub use self::{
     transaction_pattern::TransactionPattern,
 };
 use crate::btsieve::{BlockByHash, LatestBlock, Predates};
+use crate::Never;
 use bitcoin::{
     consensus::{encode::deserialize, Decodable},
-    BitcoinHash,
+    BitcoinHash, OutPoint,
 };
 use chrono::NaiveDateTime;
 use futures_core::compat::Future01CompatExt;
+use genawaiter::sync::{Co, Gen};
+use genawaiter::GeneratorState;
 use reqwest::{Client, Url};
 use std::collections::HashSet;
 
-pub async fn matching_transaction<C>(
-    mut blockchain_connector: C,
-    pattern: TransactionPattern,
+type Hash = bitcoin::BlockHash;
+type Block = bitcoin::Block;
+
+// TODO: reevaluate naming of function
+pub async fn watch_for_transaction<C>(
+    blockchain_connector: C,
     start_of_swap: NaiveDateTime,
+    from_outpoint: OutPoint,
+    unlock_script: Vec<Vec<u8>>,
 ) -> anyhow::Result<bitcoin::Transaction>
 where
-    C: LatestBlock<Block = bitcoin::Block>
-        + BlockByHash<Block = bitcoin::Block, BlockHash = bitcoin::BlockHash>
-        + Clone,
+    C: LatestBlock<Block = Block> + BlockByHash<Block = Block, BlockHash = Hash> + Clone,
 {
-    // Verify that we can successfully addPeer to the blockchain connector and check
-    // if the transaction is in the latest block.
-    let latest_block = blockchain_connector.latest_block().compat().await?;
-    if let Some(transaction) = check_block_against_pattern(&latest_block.clone(), &pattern) {
-        return Ok(transaction.clone());
+    // TODO: Refactor later (get rid of TransactionPatterns)
+    let pattern = TransactionPattern {
+        to_address: None,
+        from_outpoint: Some(from_outpoint),
+        unlock_script: Some(unlock_script),
     };
 
-    // We didn't find the transaction, now we need to do two things; keep polling
-    // for latest block so that we see transactions in new blocks and also go
-    // back up the blockchain until 'start_of_swap' i.e., look back in the
-    // past.
+    let transaction = matching_transaction(blockchain_connector, start_of_swap, |transaction| {
+        pattern.matches(&transaction)
+    })
+    .await?;
+
+    Ok(transaction)
+}
 
-    let mut prev_blockhashes: HashSet<bitcoin::BlockHash> = HashSet::new();
-    let mut missing_block_futures: Vec<_> = Vec::new();
+// TODO: reevaluate naming of function
```

Comment:
Could create some nice symmetric with the above suggestion and name this: `watch_for_outpoint_created`.

Reply:
Rationale: Sending bitcoin to an address creates an (unspent) outpoint :)